File: ai-engine/models/live_fact_check_models.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, StringConstraints


Ticker = Annotated[str, StringConstraints(strip_whitespace=True, to_upper=True, min_length=1)]
NonEmptyText = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]


class LiveFactCheckVerdict(str, Enum):
    SUPPORTED = "SUPPORTED"
    CONTRADICTED = "CONTRADICTED"
    INSUFFICIENT_EVIDENCE = "INSUFFICIENT_EVIDENCE"


class LiveFactCheckReasonCode(str, Enum):
    SUPPORTED_BY_NEWS = "supported_by_news"
    CONTRADICTED_BY_NEWS = "contradicted_by_news"
    INSUFFICIENT_RELEVANCE = "insufficient_relevance"
    EVIDENCE_NOT_SPECIFIC = "evidence_not_specific"
    RETRIEVAL_FAILED = "retrieval_failed"
    LLM_FAILED = "llm_failed"
    INVALID_LLM_RESPONSE = "invalid_llm_response"


class LiveFactCheckBatchStatus(str, Enum):
    BUFFERING = "BUFFERING"
    COMPLETED = "COMPLETED"
    DISCARDED = "DISCARDED"
    REJECTED = "REJECTED"


class LiveFactCheckSentenceRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    ticker: Ticker
    sentence: NonEmptyText
    sentence_sequence: int = Field(ge=0)
    sentence_timestamp: int = Field(gt=0)
    is_session_end: bool = False


class LiveFactCheckEvidence(BaseModel):
    model_config = ConfigDict(extra="ignore")

    doc_id: str
    title: str = ""
    snippet: str
    url: str = ""
    source: str = "unknown"
    published_at: int
    relevance_score: float = Field(ge=0.0, le=1.0)


class LiveClaimFactCheckResult(BaseModel):
    model_config = ConfigDict(extra="ignore")

    claim_id: str
    sentence_index: int = Field(ge=0, le=2)
    source_text: str
    claim: str
    claim_type: str
    verdict: LiveFactCheckVerdict
    confidence: float = Field(ge=0.0, le=1.0)
    explanation_ko: str
    reason_code: LiveFactCheckReasonCode
    evidence: list[LiveFactCheckEvidence] = Field(default_factory=list)
    retrieved_count: int = Field(default=0, ge=0)
    accepted_count: int = Field(default=0, ge=0)


class LiveFactCheckBatchResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    ticker: str
    status: LiveFactCheckBatchStatus
    buffered_count: int = Field(default=0, ge=0, le=2)
    batch_start_sequence: int | None = None
    batch_end_sequence: int | None = None
    claims: list[LiveClaimFactCheckResult] = Field(default_factory=list)
    excluded_count: int = Field(default=0, ge=0)
    extraction_llm_used: bool = False
    verification_llm_used: bool = False
    warnings: list[str] = Field(default_factory=list)
    generated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


__all__ = [
    "LiveClaimFactCheckResult",
    "LiveFactCheckBatchResponse",
    "LiveFactCheckBatchStatus",
    "LiveFactCheckEvidence",
    "LiveFactCheckReasonCode",
    "LiveFactCheckSentenceRequest",
    "LiveFactCheckVerdict",
]

File: backend/src/main/java/com/earningwhisperer/infrastructure/aiengine/LiveFactCheckModels.java
package com.earningwhisperer.infrastructure.aiengine;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

/**
 * AI Engine 실시간 팩트체크 계약(Contract 9) 의 요청/응답 모델.
 *
 * <p>AI Engine 은 응답 필드를 추가할 수 있으므로 모든 응답 레코드는
 * {@code @JsonIgnoreProperties(ignoreUnknown = true)} 로 둔다. 백엔드가 모르는 필드가
 * 하나 늘었다고 시연 도중 역직렬화가 깨지면 안 된다.
 */
public final class LiveFactCheckModels {

    private LiveFactCheckModels() {
    }

    /**
     * 문장 1개 제출 요청.
     *
     * @param sentenceTimestamp Unix Epoch Second. AI Engine 이 이 값을 기준으로 과거 N일의
     *                          뉴스 근거를 검색하므로 <b>반드시 실제 현재 시각</b>이어야 한다.
     *                          스크립트에 박힌 과거/미래 시각을 넣으면 근거가 적재되어 있어도
     *                          검색 결과가 0건이 되어 전부 INSUFFICIENT_EVIDENCE 로 떨어진다.
     */
    public record SentenceRequest(
            String ticker,
            String sentence,
            @JsonProperty("sentence_sequence") int sentenceSequence,
            @JsonProperty("sentence_timestamp") long sentenceTimestamp,
            @JsonProperty("is_session_end") boolean isSessionEnd
    ) {
    }

    /** 배치 응답. {@code status} 가 COMPLETED 일 때만 {@code claims} 가 채워진다. */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public record BatchResponse(
            String ticker,
            String status,
            @JsonProperty("buffered_count") int bufferedCount,
            @JsonProperty("batch_start_sequence") Integer batchStartSequence,
            @JsonProperty("batch_end_sequence") Integer batchEndSequence,
            List<Claim> claims,
            @JsonProperty("excluded_count") int excludedCount,
            @JsonProperty("extraction_llm_used") boolean extractionLlmUsed,
            @JsonProperty("verification_llm_used") boolean verificationLlmUsed,
            List<String> warnings
    ) {
        public boolean isCompleted() {
            return "COMPLETED".equals(status);
        }

        public boolean hasClaims() {
            return claims != null && !claims.isEmpty();
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Claim(
            @JsonProperty("claim_id") String claimId,
            @JsonProperty("sentence_index") int sentenceIndex,
            @JsonProperty("source_text") String sourceText,
            String claim,
            @JsonProperty("claim_type") String claimType,
            String verdict,
            double confidence,
            @JsonProperty("explanation_ko") String explanationKo,
            @JsonProperty("reason_code") String reasonCode,
            List<Evidence> evidence,
            @JsonProperty("retrieved_count") int retrievedCount,
            @JsonProperty("accepted_count") int acceptedCount
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Evidence(
            @JsonProperty("doc_id") String docId,
            String title,
            String snippet,
            String url,
            String source,
            @JsonProperty("published_at") long publishedAt,
            @JsonProperty("relevance_score") double relevanceScore
    ) {
    }
}
